airbyte-integrations/connectors/source-bigquery/source_bigquery/streams.py

```diff
@@ -0,0 +1,187 @@
+#
+# Copyright (c) 2024 Airbyte, Inc., all rights reserved.
+#
+
+
+import logging
+from abc import ABC
+from typing import Any, Iterable, List, Mapping, MutableMapping, Optional, Tuple
+
+import requests
+from airbyte_cdk.sources import AbstractSource
+from airbyte_cdk.sources.streams import Stream
+from airbyte_cdk.sources.streams.http import HttpStream, HttpSubStream
+from airbyte_cdk.sources.streams.http.auth import TokenAuthenticator
+from airbyte_protocol.models import SyncMode
+from airbyte_cdk.models import AirbyteCatalog, AirbyteMessage, AirbyteStateMessage, ConfiguredAirbyteCatalog
+from airbyte_cdk.logger import AirbyteLogger
+
+"""
+This file provides a stubbed example of how to use the Airbyte CDK to develop both a source connector which supports full refresh or and an
+incremental syncs from an HTTP API.
+
+The various TODOs are both implementation hints and steps - fulfilling all the TODOs should be sufficient to implement one basic and one incremental
+stream from a source. This pattern is the same one used by Airbyte internally to implement connectors.
+
+The approach here is not authoritative, and devs are free to use their own judgement.
+
+There are additional required TODOs in the files within the integration_tests folder and the spec.yaml file.
+"""
+URL_BASE: str = "https://bigquery.googleapis.com"
+    
+
+class BigqueryStream(HttpStream, ABC):
+    """
+    """ 
+    url_base = URL_BASE
+    primary_key = "id"
+    raise_on_http_errors = True
+
+    def __init__(self, stream_path: str, stream_name: str, stream_schema, **kwargs):
+        super().__init__(**kwargs)
+        self.stream_path = stream_path
+        self.stream_name = stream_name
+        self.stream_schema = stream_schema
+
+    @property
+    def name(self):
+        return self.stream_name
+
+    def get_json_schema(self) -> Mapping[str, Any]:
+        return self.stream_schema
+
+    def next_page_token(self, response: requests.Response, **kwargs) -> Optional[Mapping[str, Any]]:
+        # TODO: check if correct
+        next_page = response.json().get("offset")
+        if next_page:
+            return next_page
+        return None
+
+    def process_records(self, record) -> Iterable[Mapping[str, Any]]:
+        pass
+
+    def parse_response(self, response: requests.Response, **kwargs) -> Iterable[Mapping]:
+        records = response.json()
+        yield from self.process_records(records)
+
+    def path(self, **kwargs) -> str:
```

Comment:
It seems like `path` is always override except from `IncrementalBigqueryDatasets`. Is there a reason to define it here?

Reply:
I was thinking since in Bigquery API the endpoint to get the table schema (the one used in `BigqueryTable`class) is different than the one in which we get the table data (`BigqueryTableData` class) in the streams method of `SourceBigquery` we will have to return a `BigqueryStream` object that has its `path` set to the `BigqueryTable` path for the schema but uses an `BigqueryTableData` object for the table data.